File: Binary_Search_Trees/1050-construct-binary-search-tree-from-preorder-traversal/construct-binary-search-tree-from-preorder-traversal.py
class Solution:
    def helper(self, preorder: List[int], ind: int, limit: int) -> Optional[TreeNode]:
        if ind[0] == len(preorder) or preorder[ind[0]] > limit:
            return None

        root = TreeNode(preorder[ind[0]])
        ind[0] += 1

        root.left = self.helper(preorder, ind, root.val)
        root.right = self.helper(preorder, ind, limit)

        return root

    def bstFromPreorder(self, preorder: List[int]) -> Optional[TreeNode]:
        ind = [0]
        return self.helper(preorder, ind, float('inf'))

File: Binary_Search_Trees/1050-construct-binary-search-tree-from-preorder-traversal/construct-binary-search-tree-from-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* helper(vector<int>& preorder, int& ind, int limit){
        if(ind == preorder.size() || limit < preorder[ind]) return NULL;

        TreeNode* root = new TreeNode(preorder[ind++]);
        root->left = helper(preorder, ind, root->val);
        root->right = helper(preorder, ind, limit);

        return root;
    }
    
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        int i = 0;
        return helper(preorder, i, INT_MAX);
    }
};